File: python/firefox_bidi_client/process_manager.py
import asyncio
import os
import re
import shutil
import sys
import tempfile
from pathlib import Path
from typing import IO, Optional

from .logger import log, log_debug


class FirefoxProcessManager:
    def __init__(self):
        self._process: Optional[asyncio.subprocess.Process] = None
        self._debugging_port: int = 0
        self._profile_path: Optional[Path] = None
        self._is_temp_profile: bool = False
        self._log_file: Optional[IO[bytes]] = None
        self._stdout_task: Optional[asyncio.Task] = None
        self._stderr_task: Optional[asyncio.Task] = None

    async def launch(self, options: dict) -> int:
        log("Launching Firefox with BiDi...")

        firefox_binary = options.get("firefox_path") or self._get_default_firefox_path()
        args = self._build_firefox_args(options)

        log_debug(f"Firefox binary: {firefox_binary}")
        log_debug(f"Firefox args: {' '.join(args)}")

        env = None
        if options.get("env"):
            env = {**os.environ, **options["env"]}

        if options.get("log_file"):
            self._log_file = open(options["log_file"], "ab")

        self._process = await asyncio.create_subprocess_exec(
            firefox_binary,
            *args,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
            env=env,
        )

        self._debugging_port = await self._wait_for_bidi_port()
        log(f"Firefox launched (BiDi port: {self._debugging_port})")
        return self._debugging_port

    def _build_firefox_args(self, options: dict) -> list:
        args = ["--no-remote", "--remote-debugging-port=0"]

        if options.get("headless"):
            args.append("--headless")

        if options.get("viewport"):
            vp = options["viewport"]
            args += [f"--width={vp['width']}", f"--height={vp['height']}"]

        if options.get("profile_path"):
            self._profile_path = Path(options["profile_path"])
            self._is_temp_profile = False
        else:
            self._profile_path = Path(tempfile.mkdtemp(prefix="firefox-profile-"))
            self._is_temp_profile = True

        args += ["--profile", str(self._profile_path)]

        if options.get("args"):
            args += options["args"]

        return args

    def _get_default_firefox_path(self) -> str:
        if sys.platform == "darwin":
            return "/Applications/Firefox.app/Contents/MacOS/firefox"
        if sys.platform == "win32":
            return r"C:\Program Files\Mozilla Firefox\firefox.exe"
        return "firefox"

    async def _wait_for_bidi_port(self) -> int:
        pattern = re.compile(
            r"WebDriver BiDi listening on ws://127\.0\.0\.1:(\d+)", re.IGNORECASE
        )
        port_future: asyncio.Future = asyncio.get_event_loop().create_future()

        async def drain(stream, label: str) -> None:
            async for line in stream:
                text = line.decode("utf-8", errors="replace")
                if self._log_file:
                    self._log_file.write(line)
                log_debug(f"Firefox {label}: {text.rstrip()}")
                if not port_future.done():
                    m = pattern.search(text)
                    if m:
                        port_future.set_result(int(m.group(1)))

        self._stdout_task = asyncio.create_task(drain(self._process.stdout, "stdout"))
        self._stderr_task = asyncio.create_task(drain(self._process.stderr, "stderr"))

        try:
            return await asyncio.wait_for(asyncio.shield(port_future), timeout=10.0)
        except asyncio.TimeoutError:
            raise TimeoutError("Timeout waiting for Firefox BiDi port (10 seconds)")

    async def kill(self) -> None:
        for task in (self._stdout_task, self._stderr_task):
            if task:
                task.cancel()
        self._stdout_task = None
        self._stderr_task = None

        if self._log_file:
            self._log_file.close()
            self._log_file = None

        if self._process and self._process.returncode is None:
            log_debug("Killing Firefox process")
            try:
                self._process.terminate()
                await asyncio.wait_for(self._process.wait(), timeout=5.0)
            except asyncio.TimeoutError:
                log_debug("Force killing Firefox process")
                self._process.kill()
                await self._process.wait()

        self._process = None

        if self._is_temp_profile and self._profile_path:
            try:
                shutil.rmtree(self._profile_path, ignore_errors=True)
                log_debug(f"Cleaned up temporary profile: {self._profile_path}")
            except Exception as e:
                log_debug(f"Error cleaning up temporary profile: {e}")
            self._profile_path = None
            self._is_temp_profile = False

        self._debugging_port = 0

    def is_running(self) -> bool:
        return self._process is not None and self._process.returncode is None

    def get_port(self) -> int:
        return self._debugging_port

File: src/process-manager.ts
/**
 * Firefox Process Manager
 *
 * Handles launching Firefox with BiDi enabled and managing the process lifecycle.
 */

import { spawn, ChildProcess } from 'child_process';
import { openSync, closeSync, mkdtempSync, rmSync, writeSync } from 'fs';
import { join } from 'path';
import { tmpdir } from 'os';
import { log, logDebug } from './logger.js';
import type { FirefoxLaunchOptions } from './types.js';

export class FirefoxProcessManager {
  private process: ChildProcess | null = null;
  private debuggingPort: number = 0;
  private logFileFd: number | null = null;
  private profilePath: string | null = null;
  private isTemporaryProfile: boolean = false;

  /**
   * Launch Firefox with BiDi enabled
   * Returns the BiDi WebSocket port number
   */
  async launch(options: FirefoxLaunchOptions): Promise<number> {
    log('🚀 Launching Firefox with BiDi...');

    // Build Firefox command line arguments
    const args = this.buildFirefoxArgs(options);

    // Determine Firefox binary path
    const firefoxBinary = options.firefoxPath || this.getDefaultFirefoxPath();
    logDebug(`Firefox binary: ${firefoxBinary}`);
    logDebug(`Firefox args: ${args.join(' ')}`);

    // Set up environment variables
    const env = {
      ...process.env,
      ...(options.env || {}),
    };

    // Always use 'pipe' for stdio so we can capture BiDi port from stdout/stderr
    const stdio: any = ['ignore', 'pipe', 'pipe'];

    // Open log file if requested (we'll tee the output)
    if (options.logFile) {
      this.logFileFd = openSync(options.logFile, 'a');
      logDebug(`Logging Firefox output to: ${options.logFile}`);
    }

    // Spawn Firefox process
    this.process = spawn(firefoxBinary, args, {
      env,
      stdio,
      detached: false,
    });

    // Handle process events
    this.process.on('error', (error) => {
      logDebug(`Firefox process error: ${error.message}`);
    });

    this.process.on('exit', (code, signal) => {
      logDebug(`Firefox process exited with code ${code}, signal ${signal}`);
      this.process = null;
    });

    // Wait for Firefox to start and BiDi port to be available
    this.debuggingPort = await this.waitForBiDiPort();

    log(`✅ Firefox launched (BiDi port: ${this.debuggingPort})`);
    return this.debuggingPort;
  }

  /**
   * Build Firefox command line arguments
   */
  private buildFirefoxArgs(options: FirefoxLaunchOptions): string[] {
    const args: string[] = [];

    // Prevent Firefox from delegating to an existing running instance
    args.push('--no-remote');

    // Enable remote debugging with BiDi
    // Using port 0 tells Firefox to choose a free port
    args.push('--remote-debugging-port=0');

    // Headless mode
    if (options.headless) {
      args.push('--headless');
    }

    // Window size/viewport
    if (options.viewport) {
      args.push(`--width=${options.viewport.width}`);
      args.push(`--height=${options.viewport.height}`);
    }

    // Profile path - create temporary if not specified
    if (options.profilePath) {
      args.push('--profile', options.profilePath);
      this.profilePath = options.profilePath;
      this.isTemporaryProfile = false;
      logDebug(`Using Firefox profile: ${options.profilePath}`);
    } else {
      // Create temporary profile
      this.profilePath = mkdtempSync(join(tmpdir(), 'firefox-profile-'));
      this.isTemporaryProfile = true;
      args.push('--profile', this.profilePath);
      logDebug(`Created temporary Firefox profile: ${this.profilePath}`);
    }

    // Additional Firefox arguments
    if (options.args && options.args.length > 0) {
      args.push(...options.args);
    }

    return args;
  }

  /**
   * Get default Firefox binary path based on platform
   */
  private getDefaultFirefoxPath(): string {
    const platform = process.platform;

    switch (platform) {
      case 'darwin': // macOS
        return '/Applications/Firefox.app/Contents/MacOS/firefox';
      case 'win32': // Windows
        return 'C:\\Program Files\\Mozilla Firefox\\firefox.exe';
      case 'linux':
        return 'firefox';
      default:
        return 'firefox';
    }
  }

  /**
   * Wait for Firefox to start and discover the BiDi port from stdout/stderr
   * Firefox outputs: "WebDriver BiDi listening on ws://127.0.0.1:PORT"
   */
  private async waitForBiDiPort(): Promise<number> {
    return new Promise((resolve, reject) => {
      const timeout = setTimeout(() => {
        reject(new Error('Timeout waiting for Firefox BiDi port (10 seconds)'));
      }, 10000);

      let stdoutBuffer = '';
      let stderrBuffer = '';

      const checkForPort = (data: string): number | null => {
        // Look for: "WebDriver BiDi listening on ws://127.0.0.1:9222"
        const match = data.match(/WebDriver BiDi listening on ws:\/\/127\.0\.0\.1:(\d+)/i);
        if (match && match[1]) {
          return parseInt(match[1], 10);
        }
        return null;
      };

      const stdoutHandler = (data: Buffer) => {
        const text = data.toString();
        stdoutBuffer += text;

        // Write to log file if configured
        if (this.logFileFd !== null) {
          try {
            writeSync(this.logFileFd, data);
          } catch (error) {
            logDebug(`Error writing to log file: ${error}`);
          }
        }

        // Also log to debug
        logDebug(`Firefox stdout: ${text.trim()}`);

        const port = checkForPort(stdoutBuffer);
        if (port) {
          clearTimeout(timeout);
          logDebug(`Found BiDi port from stdout: ${port}`);

          // Keep listeners for logging but don't wait for more data
          resolve(port);
        }
      };

      const stderrHandler = (data: Buffer) => {
        const text = data.toString();
        stderrBuffer += text;

        // Write to log file if configured
        if (this.logFileFd !== null) {
          try {
            writeSync(this.logFileFd, data);
          } catch (error) {
            logDebug(`Error writing to log file: ${error}`);
          }
        }

        // Also log to debug
        logDebug(`Firefox stderr: ${text.trim()}`);

        const port = checkForPort(stderrBuffer);
        if (port) {
          clearTimeout(timeout);
          logDebug(`Found BiDi port from stderr: ${port}`);

          // Keep listeners for logging but don't wait for more data
          resolve(port);
        }
      };

      // Listen to stdout and stderr
      if (this.process) {
        this.process.stdout?.on('data', stdoutHandler);
        this.process.stderr?.on('data', stderrHandler);

        // Check if process exits before we find the port
        this.process.once('exit', (code) => {
          clearTimeout(timeout);
          reject(new Error(`Firefox process exited unexpectedly with code ${code}`));
        });
      } else {
        clearTimeout(timeout);
        reject(new Error('No Firefox process'));
      }
    });
  }

  /**
   * Kill Firefox process
   */
  async kill(): Promise<void> {
    if (this.process) {
      logDebug('Killing Firefox process');

      // Close log file if open
      if (this.logFileFd !== null) {
        try {
          closeSync(this.logFileFd);
        } catch (error) {
          logDebug(`Error closing log file: ${error}`);
        }
        this.logFileFd = null;
      }

      // Kill process
      if (this.process.exitCode === null) {
        this.process.kill('SIGTERM');

        // Wait up to 5 seconds for graceful shutdown
        const killTimeout = setTimeout(() => {
          if (this.process && this.process.exitCode === null) {
            logDebug('Force killing Firefox process');
            this.process.kill('SIGKILL');
          }
        }, 5000);

        // Wait for process to exit
        await new Promise<void>((resolve) => {
          if (!this.process || this.process.exitCode !== null) {
            clearTimeout(killTimeout);
            resolve();
            return;
          }

          this.process.once('exit', () => {
            clearTimeout(killTimeout);
            resolve();
          });
        });
      }

      this.process = null;
    }

    // Clean up temporary profile
    if (this.isTemporaryProfile && this.profilePath) {
      try {
        logDebug(`Cleaning up temporary profile: ${this.profilePath}`);
        rmSync(this.profilePath, { recursive: true, force: true });
      } catch (error) {
        logDebug(`Error cleaning up temporary profile: ${error}`);
      }
      this.profilePath = null;
      this.isTemporaryProfile = false;
    }

    this.debuggingPort = 0;
  }

  /**
   * Check if Firefox process is running
   */
  isRunning(): boolean {
    return this.process !== null && this.process.exitCode === null;
  }

  /**
   * Get the BiDi port number
   */
  getPort(): number {
    return this.debuggingPort;
  }
}
